### PR TITLE
First pass at adding optical duplicate flagging/removal into MarkDupes

### PR DESCRIPTION
**Pull request by tfenne**
First pass at some changes to MarkDuplicates to allow optical (co-localized?) duplicates to be annotated in the output file and/or excluded from the output file.  So far support is only included in MarkDuplicates, though with a little help from @nh13 or @bradtaylor I'd be happy to extend support to MarkDuplicatesWithMateCigar also.

@nh13, @yfarjoun : Could you take a first look at this please and let me know what you think of the direction I've taken?  I'm not ready (yet) for a detailed review, just looking to get some extra eyes on this before I sink more time into writing tests and shining it up.  Thanks!

### CODE REVIEW

**Comment by tfenne** [comment]
@yfarjoun I've updated this to:
1. Use the `DT` optional tag
2. Provide options for no-tagging, tag only optical duplicate, tag all duplicates
3. Add more to the documentation

Since I know you're interested in this kind of thing, would you be up for a review?  There's no huge rush to merge, but I'd like to get any major problems/changes identified and fixed quickly.

**Comment by yfarjoun** [comment]
I'd be happy to take a look.

**Comment by sooheelee** [comment]
When are you planning to make this public? I've been working on an [updated  MarkDuplicates tutorial](http://gatkforums.broadinstitute.org/gatk/discussion/6747/how-to-flag-duplicates-with-markduplicates-or-markduplicateswithmatecigar#latest) that includes MarkDuplicatesWithMateCigar.

**Comment by tfenne** [comment]
@sooheelee It's in public now, but I assume you mean when am I planning on merging it?  I'm happy with it in it's current form, so really it will depend on @yfarjoun's availability to review, and what he has to say.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 1]
```diff
@@ -109,9 +109,19 @@
     @Option(doc = "Read two barcode SAM tag (ex. BX for 10X Genomics)", optional = true)
     public String READ_TWO_BARCODE_TAG = null;
 
+    @Option(doc = "If true remove optical duplicates from the output even if REMOVE_DUPLICATES is false.")
+    public boolean REMOVE_OPTICAL_DUPLICATES = false;
+
+    @Option(doc = "If set write a tag with this name and OPTICAL_DUPLICATE_TAG_VALUE on to each read that is an optical duplicate.", optional=true)
+    public String OPTICAL_DUPLICATE_TAG_NAME = null;
+
+    @Option(doc = "If true do not write duplicates to the output file instead of writing them with appropriate flags set.")
```
Comment (eitanbanks) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: I don't think this description is correct.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 2]
```diff
@@ -109,9 +109,19 @@
     @Option(doc = "Read two barcode SAM tag (ex. BX for 10X Genomics)", optional = true)
     public String READ_TWO_BARCODE_TAG = null;
 
+    @Option(doc = "If true remove optical duplicates from the output even if REMOVE_DUPLICATES is false.")
+    public boolean REMOVE_OPTICAL_DUPLICATES = false;
```
Comment (eitanbanks) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: Might want to mention that if REMOVE_DUPLICATES is set to true then it overrides this being set to false?

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 3]
```diff
@@ -72,9 +72,12 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
      * unmodified a copy of the list should be passed to this method.
      *
      * @param list a list of reads that are determined to be duplicates of one another
+     * @param keeper a single PhysicalLocation that is the one being kept as non-duplicate, and thus should never be
+     *               annotated as an optical duplicate. May in some cases be null, or a PhysicalLocation not
+     *               contained within the list!
```
Comment (eitanbanks) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: Could you please explain why this is necessary?  Is it possible to have multiple optical duplicate sets within a given duplicate set?

**Comment by eitanbanks** [comment]
Review complete.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 4]
```diff
@@ -109,9 +109,19 @@
...
+    @Option(doc = "If set write a tag with this name and OPTICAL_DUPLICATE_TAG_VALUE on to each read that is an optical duplicate.", optional=true)
+    public String OPTICAL_DUPLICATE_TAG_NAME = null;
+
+    @Option(doc = "If true do not write duplicates to the output file instead of writing them with appropriate flags set.")
+    public String OPTICAL_DUPLICATE_TAG_VALUE = "+";
+
     private SortingCollection<ReadEndsForMarkDuplicates> pairSort;
     private SortingCollection<ReadEndsForMarkDuplicates> fragSort;
     private SortingLongCollection duplicateIndexes;
+    private SortingLongCollection opticalDuplicateIndices;
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: even if both of the spellings of the plural of index are correct, I suggest we use one.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 5]
```diff
@@ -109,9 +109,19 @@
     @Option(doc = "Read two barcode SAM tag (ex. BX for 10X Genomics)", optional = true)
     public String READ_TWO_BARCODE_TAG = null;
 
+    @Option(doc = "If true remove optical duplicates from the output even if REMOVE_DUPLICATES is false.")
+    public boolean REMOVE_OPTICAL_DUPLICATES = false;
+
+    @Option(doc = "If set write a tag with this name and OPTICAL_DUPLICATE_TAG_VALUE on to each read that is an optical duplicate.", optional=true)
+    public String OPTICAL_DUPLICATE_TAG_NAME = null;
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: according to https://github.com/samtools/hts-specs/issues/121 this should be hard-coded to DT, a new tag to be added to SAMUtils.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 6]
```diff
@@ -109,9 +109,19 @@
...
+
+    @Option(doc = "If true do not write duplicates to the output file instead of writing them with appropriate flags set.")
+    public String OPTICAL_DUPLICATE_TAG_VALUE = "+";
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: should this default to "optical" rather than '+'? should this even be an option?

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 7]
```diff
@@ -226,15 +237,22 @@ protected int doWork() {
...
+            final boolean isOpticalDuplicate = recordInFileIndex == nextOpticalDuplicateIndex;
+            if (isOpticalDuplicate) {
+                if (this.OPTICAL_DUPLICATE_TAG_NAME != null) rec.setAttribute(this.OPTICAL_DUPLICATE_TAG_NAME, this.OPTICAL_DUPLICATE_TAG_VALUE);
+                nextOpticalDuplicateIndex = this.opticalDuplicateIndices.hasNext() ? this.opticalDuplicateIndices.next() : -1;
             }
+
+            // Output the record if desired and bump the record index
+            recordInFileIndex++;
+            if (this.REMOVE_DUPLICATES && rec.getDuplicateReadFlag()) continue;
+            if (this.REMOVE_OPTICAL_DUPLICATES && isOpticalDuplicate) continue;
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: if you move the index bump before the if(isBoolean), you will be able to put this continue inside the if statement. it's slightly simpler since it will put all the optical stuff in the same place.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 8]
```diff
@@ -460,10 +478,13 @@ private ReadEndsForMarkDuplicates buildReadEnds(final SAMFileHeader header, fina
      *
      * @return an array with an ordered list of indexes into the source file
      */
-    private void generateDuplicateIndexes(final boolean useBarcodes) {
+    private void generateDuplicateIndexes(final boolean useBarcodes, final boolean indexOpticalDuplicates) {
         // Keep this number from getting too large even if there is a huge heap.
-        final int maxInMemory = (int) Math.min((Runtime.getRuntime().maxMemory() * 0.25) / SortingLongCollection.SIZEOF,
-                (double) (Integer.MAX_VALUE - 5));
+        int maxInMemory = (int) Math.min((Runtime.getRuntime().maxMemory() * 0.25) / SortingLongCollection.SIZEOF, (double) (Integer.MAX_VALUE - 5));
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: I think it will be clearer if maxInmemory remains final and you use maxinMemory/2 as the input for the opticalDupIndex. BTW, can you comment on why you need half the amount?

Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: Oh. sorry. I get it. you're using half the available memory for the regular indices and half for the opticals...

nm. ignore previous comment about final. just add a comment that explains this. thanks.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 9]
```diff
@@ -460,10 +478,13 @@ private ReadEndsForMarkDuplicates buildReadEnds(final SAMFileHeader header, fina
...
+        if (indexOpticalDuplicates) {
+            maxInMemory /= 2;
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: add a log explaining that keeping track of optical dups will mean more IO.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java (hunk 3 above)
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: yes. imagine that you have two PCR copies, A and B and each has optical dups A' and B'. any of these could be keeper, so you want to make sure you do not mark that one as an optical dup.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 10]
```diff
@@ -96,13 +99,14 @@ public int compare(final PhysicalLocation lhs, final PhysicalLocation rhs) {
...
-                if (opticalDuplicateFlags[j]) continue;
+                if ((rhs == keeper && opticalDuplicateFlags[i]) || opticalDuplicateFlags[j]) continue;
                 if (lhs.getReadGroup() != rhs.getReadGroup()) continue outer;
                 if (lhs.getTile() != rhs.getTile()) continue outer;
                 if (rhs.getX() > lhs.getX() + this.opticalDuplicatePixelDistance) continue outer;
 
                 if (Math.abs(lhs.getY() - rhs.getY()) <= this.opticalDuplicatePixelDistance) {
-                    opticalDuplicateFlags[j] = true;
+                    if (rhs == keeper) opticalDuplicateFlags[i] = true;
+                    else               opticalDuplicateFlags[j] = true;
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: the logic here seems right. but it's a bit messy....would be good to have another pair of eyes on it and also specific tests.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 11]
```diff
@@ -72,9 +72,12 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
      * unmodified a copy of the list should be passed to this method.
      *
      * @param list a list of reads that are determined to be duplicates of one another
+     * @param keeper a single PhysicalLocation that is the one being kept as non-duplicate, and thus should never be
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: I keep thinking of sports analogies for `keeper`, how about `representative` instead like in htsjdk's [DuplicateSet](https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/DuplicateSet.java)?

**Comment by yfarjoun** [comment]
@tfenne I think it's a good direction. I think it would be great value having these tags in the reads. 

With the newest sequencers 'short' doesn't cut it anymore for physical position, we need a full int to avoid overflow. I've run into this problem when writing the PositionBasedDownsampling, which is why I implemented PhysicalPositionInt. Do you think this would be a good time to switch everything over to ints or do you think that the memory requirements would be too onerous? an alternative would be to allow for overflow and take it into consideration when calculating the pixel distance (overflow is blatantly ignored currently) what do you think?

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 12]
```diff
@@ -96,13 +99,14 @@ public int compare(final PhysicalLocation lhs, final PhysicalLocation rhs) {
             for (int j = i + 1; j < length; ++j) {
                 final PhysicalLocation rhs = list.get(j);
 
-                if (opticalDuplicateFlags[j]) continue;
+                if ((rhs == keeper && opticalDuplicateFlags[i]) || opticalDuplicateFlags[j]) continue;
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: `(rhs == keeper && opticalDuplicateFlags[i])` can actually be moved into the outer loop.

Comment (nh13) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: Not true since `opticalDuplicateFlags[i]` may be set below.  Wow this is tough to follow.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 13]
```diff
@@ -96,13 +99,14 @@ public int compare(final PhysicalLocation lhs, final PhysicalLocation rhs) {
...
                 if (lhs.getReadGroup() != rhs.getReadGroup()) continue outer;
                 if (lhs.getTile() != rhs.getTile()) continue outer;
                 if (rhs.getX() > lhs.getX() + this.opticalDuplicatePixelDistance) continue outer;
 
                 if (Math.abs(lhs.getY() - rhs.getY()) <= this.opticalDuplicatePixelDistance) {
-                    opticalDuplicateFlags[j] = true;
+                    if (rhs == keeper) opticalDuplicateFlags[i] = true;
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: can you break here since now `(rhs == keeper && opticalDuplicateFlags[i])` will always be true, which causes the continue statement on line `102` to always be invoked?

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java (hunk 3 above)
Comment (eitanbanks) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: Got it.  I know you guys have said this below, but I think there needs to be a lot of comments in the code explaining what is happening below.

Comment (tfenne) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: It's because, to do the optical duplicate detection, you need to see all the locations, including the read pair that is being kept as non-duplicate.  I.e. if you have two read pairs with same alignment positions, you need both to determine if one is an optical duplicate of the other, but you should only mark one of the with an optical duplicate tag.  And ideally *not* the one that isn't being marked as a duplicate.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java (hunk 4 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: I didn't even notice that I did that.  Sorry.

**Comment by tfenne** [comment]
Hey @eitanbanks and @yfarjoun, I'm really sorry but I just realized that I messed up.  I made some substantial changes go this before I re-pinged for review, and I was sure I had pushed those changes up, but now I see that my push failed and I didn't see it at the time.

I'm going to go through and apply any changes you guys suggested that still make sense, then rebase and push again.  I'll try and highlight the differences when I do that.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java (hunk 5 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: Agreed.  This was changed but not pushed.  Sorry.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java (hunk 6 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: Ditto.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java (hunk 11 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: I don't like `representative` because the name alone is not sufficient to convey that this _must_ be the records that is considered non-duplicate amongst the set.  How about `nonDuplicate` or `retained` instead?

**Comment by tfenne** [comment]
@yfarjoun Regarding shorts vs. ints, it's probably time we just pushed everything over to ints, or (shudder) perhaps even floats?  That said, I don't really want to take that on as part of this PR.

The only other option I can think of, just to throw it out there, would be to _parse_ the locations out as ints, and then repeatedly divide by 10 and round until they fit in a short.  I say this because shorts should be enough to store a location within a (square) gigapixel image (2^15 * 2^15), and I think this is only an issue because of Illumina multiplying up the pixel coordinates to do sub-pixel stuff.  If we only need pixel resolution, we could push them back down?  Just a thought.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java (hunk 13 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: @yfarjoun  is right that this is convoluted now.  I'll add a few tests, and then try what you suggest.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 14]
```diff
@@ -226,15 +254,28 @@ protected int doWork() {
                     rec.setDuplicateReadFlag(false);
                 }
             }
-            recordInFileIndex++;
 
-            if (!this.REMOVE_DUPLICATES || !rec.getDuplicateReadFlag()) {
-                if (PROGRAM_RECORD_ID != null) {
-                    rec.setAttribute(SAMTag.PG.name(), chainedPgIds.get(rec.getStringAttribute(SAMTag.PG.name())));
+            // Manage the flagging of optical/sequencing duplicates
```
Comment (tfenne) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: @yfarjoun, @eitanbanks : this section is probably the biggest change (along with the change in CLP args), as it now will optionally tag no duplicates, optical only, or all duplicates with a tag.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 15]
```diff
@@ -67,45 +67,36 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
...
-                if (retval == 0) retval = lhs.getTile() - rhs.getTile();
-                if (retval == 0) retval = lhs.getX() - rhs.getX();
-                if (retval == 0) retval = lhs.getY() - rhs.getY();
-                return retval;
-            }
-        });
-
-        outer:
-        for (int i = 0; i < length; ++i) {
+        for (int i=0; i<length; ++i) {
```
Comment (tfenne) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: @yfarjoun , @eitanbanks , @nh13 : this method is now *much* simpler, though likely less efficient for large piles of duplicates.  None of us (me included) caught the fact that the method returns an array of flags that is index-synced to the list of PhysicalLocations - but only the post-sorted version.  That seems really nasty and error-prone, so I removed the sort and simplified the code.

**Comment by tfenne** [comment]
Ready for round 2.  Sorry again about having screwed up the push before the review.

**Comment by tfenne** [comment]
@yfarjoun Any chance you can take another look soon?

**Comment by yfarjoun** [comment]
sure. thanks for the ping.

**Comment by yfarjoun** [comment]
back to you @tfenne. Please add the test-case that I highlighted in the comments (or a similar one)

**Comment by tfenne** [comment]
@yfarjoun Good catch on the clique/keeper test case. I've added that exact test case and fixed the code to make it work.  I think the resulting code is, largely, clearer.  Hit your other fixes too.  Do you want to take a quick look at the revamped `findOpticalDuplicates()` method or are you happy with all the tests passing (including the new one)?

**Comment by tfenne** [comment]
Looks like the latest change broke a bunch of other MarkDuplicates tests, I'm investigating now.

**Comment by tfenne** [comment]
Ok, one more try - this most recent push should have all the tests passing again.

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 16]
```diff
@@ -65,47 +65,76 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
...
+        if (keeper != null && keeper.hasLocation()) {
+            for (final PhysicalLocation loc : list) {
+                if (loc == keeper) return keeper;
+            }
+        }
+        return null;
+    }
+
+    /** Simple method to test whether two physical locations are close enough to each other to be deemed optical dupes. */
+    private boolean closeEnough(final PhysicalLocation lhs, final PhysicalLocation rhs, final int distance) {
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: static

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 17]
```diff
@@ -65,47 +65,76 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
...
+                    opticalDuplicateFlags[index] = true;
                 }
             }
         }
+        
         return opticalDuplicateFlags;
     }
+
+    /** Returns the keeper if it is contained within the list and has location information, otherwise null. */
+    private PhysicalLocation keeperOrNull(final List<? extends PhysicalLocation> list, final PhysicalLocation keeper) {
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: static

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 18]
```diff
@@ -65,47 +65,76 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
...
-                if (opticalDuplicateFlags[j]) continue;
-                if (lhs.getReadGroup() != rhs.getReadGroup()) continue outer;
-                if (lhs.getTile() != rhs.getTile()) continue outer;
-                if (rhs.getX() > lhs.getX() + this.opticalDuplicatePixelDistance) continue outer;
-
-                if (Math.abs(lhs.getY() - rhs.getY()) <= this.opticalDuplicatePixelDistance) {
-                    opticalDuplicateFlags[j] = true;
+                if (closeEnough(lhs, rhs, distance)) {
+                    // At this point we want to mark either lhs or rhs as duplicate. Generally we want to mark
+                    // rhs, unless rhs is already marked.  The only way rhs can already be marked is if it's a
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: s/is if/as if/ in comment.

Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: I agree with the code, but not with the comment.

In the case of no keeper and x positions 10,20,30 with distance 25, by the time i=1, j=2 both flags will be set to true, so it's not the "only way". I'd say "The only way rhs can already be marked is if it's a duplicate of a "previous" read (perhaps the keeper) so in that case it's safe to mark the lhs, and so we do."

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java [hunk 19]
```diff
@@ -65,47 +65,76 @@ public OpticalDuplicateFinder(final String readNameRegex, final int opticalDupli
...
                 }
             }
         }
+        
         return opticalDuplicateFlags;
     }
+
+    /** Returns the keeper if it is contained within the list and has location information, otherwise null. */
+    private PhysicalLocation keeperOrNull(final List<? extends PhysicalLocation> list, final PhysicalLocation keeper) {
+        if (keeper != null && keeper.hasLocation()) {
```
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: please test this (add a test where keeper isn't part of the list, and isn't null)

**Comment by yfarjoun** [comment]
ok. only minor comments this time round. some comment changes and another test. what are the next steps here? squash and merge?

File: src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java (hunk 17 above)
Comment (tfenne) [inline src/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java]: I agree that this could be static, but I see no advantage to making it static - it's only referenced from a non-static context.

**Comment by tfenne** [comment]
Nulls were being tested already.  I added two more tests:
1. Just randomly shuffles a list a hundred times and randomly selects the keeper also, and then ensures that it always gets the right number of optical duplicates back.
2. Tests with a keeper that's not in the list

And also fixed that comment that was out of date with the latest changes.  I've squashed - I think this is ready to merge.

**Comment by tfenne** [comment]
@yfarjoun All tests are passing - want to do the honours?

**Comment by yfarjoun** [comment]
One more thing (from me) could you extract the string literals to constants (both the sam tag and the enum strings)?

**Comment by tfenne** [comment]
@yfarjoun Those strings have been constant-ized.  Merging now.
